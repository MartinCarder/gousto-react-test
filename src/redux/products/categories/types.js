export const PRODUCTS_CATEGORY_FETCH = "PRODUCTS/CATEGORY_FETCH";
export const PRODUCTS_CATEGORY_FETCH_SUCCESS = "PRODUCTS/CATEGORY_FETCH_SUCCESS";
export const PRODUCTS_CATEGORY_FETCH_FAILED = "PRODUCTS/CATEGORY_FETCH_FAILED";

export const status = {
  LOADING: "LOADING",
  LOADED: "LOADED",
  FAILED: "FAILED"
};
