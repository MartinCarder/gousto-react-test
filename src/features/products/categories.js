import React from "react";
import { NavLink } from "react-router-dom";
import PropTypes from "prop-types";

const CategoryPresentation = ({ categories }) => (
  <div>
    {categories.map(category => (
      <div key={category.id}>
        <NavLink to={`/products/${category.id}`}>{category.title}</NavLink>
      </div>
    ))}
  </div>
);

CategoryPresentation.propTypes = {
  categories: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired
    })
  ).isRequired
};

export default CategoryPresentation;
