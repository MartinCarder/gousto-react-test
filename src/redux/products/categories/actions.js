import {
  PRODUCTS_CATEGORY_FETCH,
  PRODUCTS_CATEGORY_FETCH_FAILED,
  PRODUCTS_CATEGORY_FETCH_SUCCESS
} from "./types";

export const categoryFetch = () => ({
  type: PRODUCTS_CATEGORY_FETCH
});

export const categoryFetchSuccess = data => ({
  type: PRODUCTS_CATEGORY_FETCH_SUCCESS,
  data
});

export const categoryFetchFailed = () => ({
  type: PRODUCTS_CATEGORY_FETCH_FAILED
});
