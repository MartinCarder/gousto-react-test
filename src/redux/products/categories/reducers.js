import {
  PRODUCTS_CATEGORY_FETCH,
  PRODUCTS_CATEGORY_FETCH_FAILED,
  PRODUCTS_CATEGORY_FETCH_SUCCESS,
  status
} from "./types";

export const initialState = {
  status: status.LOADING,
  categories: []
};

export default (state = initialState, action) => {
  switch (action.type) {
    case PRODUCTS_CATEGORY_FETCH:
      return {
        ...state,
        status: status.LOADING
      };
    case PRODUCTS_CATEGORY_FETCH_FAILED:
      return {
        ...state,
        status: status.FAILED
      };
    case PRODUCTS_CATEGORY_FETCH_SUCCESS:
      return {
        ...state,
        status: status.LOADED,
        categories: [...action.data]
      };
    default:
      return state;
  }
};
