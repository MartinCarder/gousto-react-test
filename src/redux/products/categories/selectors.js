export const getCategories = state => state.categories.categories;

export const getCategoriesStatus = state => state.categories.status;
