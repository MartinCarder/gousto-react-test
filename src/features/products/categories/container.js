import React, { useEffect } from "react";
import { connect } from "react-redux";

import { categoryFetch } from "redux/products/categories/actions";

export const Product = ({ getCategory }) => {
  useEffect(() => {
    getCategory();
  }, []);
  return <div>Product</div>;
};

export const mapDispatchToProps = {
  getCategory: categoryFetch
};

export default connect(
  null,
  mapDispatchToProps
)(Product);
