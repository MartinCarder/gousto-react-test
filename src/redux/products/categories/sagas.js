import { all, call, fork, put, takeEvery } from "redux-saga/effects";
import { apiRequest } from "redux/utils/services";
import { categoryFetchSuccess, categoryFetchFailed } from "./actions";
import { PRODUCTS_CATEGORY_FETCH } from "./types";

export function* getCategories() {
  try {
    const res = yield call(apiRequest, "get", "products/v2.0/categories");

    if (res.status === "ok") {
      yield put(categoryFetchSuccess(res.data));
    } else {
      yield put(categoryFetchFailed());
    }
  } catch (err) {
    yield put(categoryFetchFailed());
  }
}

export function* watchCategoriesFetch() {
  yield takeEvery(PRODUCTS_CATEGORY_FETCH, getCategories);
}

export default function* productSagas() {
  yield all([fork(watchCategoriesFetch)]);
}
