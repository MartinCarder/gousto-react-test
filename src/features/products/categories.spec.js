import React from "react";
import { shallow } from "enzyme";

import CategoryList from "./categories";
import { CategoryLink } from "./categories.styled";

describe("CategoryList", () => {
  let wrapper;
  const props = {
    categories: [
      {
        id: "1",
        title: "Title 1"
      },
      {
        id: "2",
        title: "Title 2"
      }
    ]
  };

  beforeEach(() => {
    wrapper = shallow(<CategoryList {...props} />);
  });

  it("Renders and matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("Displays correct number of CategoryLink", () => {
    expect(wrapper.find(CategoryLink).length).toEqual(2);
  });

  it("CategoryLink display correct title", () => {
    expect(
      wrapper
        .find(CategoryLink)
        .at(0)
        .text()
    ).toEqual(props.categories[0].title);
  });

  it("CategoryLink uses correct link", () => {
    expect(
      wrapper
        .find(CategoryLink)
        .at(0)
        .props()
        .to
    ).toEqual(`/products/${props.categories[0].id}`);
  });
});
