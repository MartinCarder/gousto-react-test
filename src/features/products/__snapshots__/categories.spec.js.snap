// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryList Renders and matches snapshot 1`] = `
<div>
  <Styled(NavLink)
    key="1"
    to="/products/1"
  >
    Title 1
  </Styled(NavLink)>
  <Styled(NavLink)
    key="2"
    to="/products/2"
  >
    Title 2
  </Styled(NavLink)>
</div>
`;
